core/src/test/resources/org/apache/calcite/test/TopDownOptTest.xml

```diff
@@ -35,6 +35,46 @@ EnumerableLimit(fetch=[5])
   EnumerableSortedAggregate(group=[{3}], EXPR$1=[COUNT()])
     EnumerableSort(sort0=[$3], dir0=[DESC-nulls-last])
       EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+    </TestCase>
+    <TestCase name="testSortProject">
+        <Resource name="sql">
+            <![CDATA[select mgr from sales.emp order by mgr desc nulls last]]>
+        </Resource>
+        <Resource name="planBefore">
+            <![CDATA[
+LogicalSort(sort0=[$0], dir0=[DESC-nulls-last])
+  LogicalProject(MGR=[$3])
+    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planAfter">
+            <![CDATA[
+EnumerableProject(MGR=[$3])
+  EnumerableSort(sort0=[$3], dir0=[DESC-nulls-last])
+    EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+    </TestCase>
+    <TestCase name="testSortFilter">
+        <Resource name="sql">
+            <![CDATA[select mgr from sales.emp where deptno > 10 order by mgr desc]]>
+        </Resource>
+        <Resource name="planBefore">
+            <![CDATA[
+LogicalSort(sort0=[$0], dir0=[DESC])
+  LogicalProject(MGR=[$3])
+    LogicalFilter(condition=[>($7, 10)])
+      LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planAfter">
+            <![CDATA[
+EnumerableProject(MGR=[$3])
+  EnumerableSort(sort0=[$3], dir0=[DESC])
+    EnumerableFilter(condition=[>($7, 10)])
+      EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
```

Comment:
I setup debugger and conformed that there was a `EnumerableSort` enforced for `EnumerableFilter`, however, I wasn't sure why planner always think that this plan is cheapest. In fact, I was hoping to see the following plan:

```
EnumerableProject(MGR=[$3])
  EnumerableFilter(condition=[>($7, 10)])
    EnumerableSort(sort0=[$3], dir0=[DESC])
      EnumerableTableScan(table=[[CATALOG, SALES, EMP]])
```

Reply:
That makes sense, because it is cheaper.
Let's think it another way.
```
select a, b, c 
  from (select a, b, max(c) c from foo group by a, b) as t 
  where a+b+c>1 
order by b desc, a desc;
```

Reply:
In many cases, Sort should always be put on top of Filter. But in this case, we might get better plan by pushing down sort requirement past Filter.

Reply:
Trait enforcement doesn't mean the relnode generated by passing through required traits is definitely best, it just adds an alternative.

Reply:
That's a good point that filter first then sort after, since filter will reduce volume of data thus reduce cost of sort. (not absolute but make senses).

Reply:
Can you use the query I provided as the test case for Filter to replace the one you added?

Reply:
Done